lib/workers.js

```diff
@@ -1,120 +1,93 @@
+          pc.ensureChild();
+          load.startJob('python_worker_idle', pc.uuid);
+          return pc;
+        },
+        destroy: async (pc) => {
+          pc.done();
+        },
+      },
+      {
+        min: numWorkers,
+        max: numWorkers,
+      }
+    );
+  },
+
+  async finishAsync() {
+    debug('finish(): waiting for pool to drain');
+    await pool.drain();
+    await pool.clear();
+    debug('finish(): pool finished draining');
   },
 
   finish(callback) {
-    debug('finish()');
-    if (!config.useWorkers) {
-      callback(null);
-      return;
-    }
-    const testFinished = () => {
-      if (availablePythonCallers.length < pythonCallers.length) {
-        // keep waiting for all pythonCallers to be returned
-        debug('finish(): waiting for availablePythonCallers');
-        setTimeout(testFinished, 100);
-      } else {
-        debug('finish(): calling done() on all pythonCallers');
-        for (let i = 0; i < pythonCallers.length; i++) {
-          pythonCallers[i].done();
-        }
-        pythonCallers = [];
-        availablePythonCallers = [];
-        return callback(null);
-      }
-    };
-    testFinished();
+    util.callbackify(module.exports.finishAsync)(callback);
   },
 
   getPythonCaller(callback) {
+    const jobUuid = uuidv4();
+    load.startJob('python_callback_waiting', jobUuid);
+
     debug('getPythonCaller()');
-    if (!config.useWorkers) {
-      const pc = new codeCaller.PythonCaller();
-      load.startJob('python_worker_active', pc.uuid);
-      callback(null, pc);
-      return;
+    if (!config.workerUseQueue) {
+      debug(`getPythonCaller(): no workers available, waiting to error`);
+      setTimeout(() => {
+        debug(`getPythonCaller(): no workers available, reporting error`);
+        callback(new Error('Server is overloaded. Please try again.'));
+      }, config.workerOverloadDelayMS);
     }
-    if (pythonCallers.length === 0) return callback(new Error('no PythonCallers initialized'));
-    if (availablePythonCallers.length > 0) {
-      const pc = availablePythonCallers[0];
-      availablePythonCallers = availablePythonCallers.slice(1);
-      load.endJob('python_worker_idle', pc.uuid);
-      debug(`getPythonCaller(): got ${pc.number}`);
-      load.startJob('python_worker_active', pc.uuid);
-      callback(null, pc);
-    } else {
-      if (config.workerUseQueue) {
-        debug(`getPythonCaller(): adding to waitingCallbacks`);
-        callback.__load_uuid = uuidv4();
-        load.startJob('python_callback_waiting', callback.__load_uuid);
-        waitingCallbacks.push(callback);
-      } else {
-        debug(`getPythonCaller(): no workers available, waiting to error`);
-        setTimeout(() => {
-          debug(`getPythonCaller(): no workers available, reporting error`);
-          callback(new Error('Server is overloaded. Please try again.'));
-        }, config.workerOverloadDelayMS);
+
+    pool.acquire().then(
```

Comment:
It is by default FIFO, yes. Check out the `fifo` option in the docs - it defaults to true, but can be changed to false to make it behave like a stack.

Reply:
I think the `fifo` option is referring to the FIFO behavior of the worker pool, not the waiting acquire queue? That is, `fifo: true` means that the oldest released worker will be the next one used (in the case where we have more workers than acquirers), whereas I'm thinking about the ordering of which waiting acquire call is given the next available worker (in the case where we have more acquirers than workers).

Reply:
Ahhhh, ok, you're reading this more carefully than I am! I'm going to assume that the queueing of requested resources is a queue because who in their right mind would build that as a stack???

Reply:
Yeah, I agree that it would be insane to use anything except a FIFO queue for the acquire queue. I'm happy to assume they are being sensible!